Initial neo-vm support

osmirnov opened the pull request:
I've started the implementation of NeoVM support. My goal not only make it running but generalize the API and adapt HyperVM later.

osmirnov committed:
initial attempt to persist invoc tx

osmirnov committed:
Merge remote-tracking branch 'upstream/development' into feat-invoc-tx

osmirnov committed:
added message container

osmirnov committed:
removed message provider

osmirnov committed:
initial NeoVM support

osmirnov committed:
moved StateMachine and StateReader from NEO

osmirnov committed:
adjusted namespaces

osmirnov committed:
drop explicit deps from NEO.VM

osmirnov committed:
update NuGet

shargon commented on src/NeoSharp.VM.NeoVM/StackItemType.cs:
```diff
@@ -0,0 +1,13 @@
+﻿namespace NeoSharp.VM.NeoVM
+{
+    internal enum StackItemType : byte
```
This enum is already on the nuget Package ( NeoSharp.VM.Interop )

shargon commented on src/NeoSharp.VM.NeoVM/StackAccessor.cs:
```diff
@@ -0,0 +1,72 @@
+{
+    public class StackAccessor : IStackAccessor
+    {
+        private readonly RandomAccessStack<StackItem> _stack;
```
We need to dispose the items when is not used, native VM require this

shargon commented on src/NeoSharp.Core/SmartContract/TriggerType.cs:
```diff
@@ -0,0 +1,22 @@
+﻿namespace NeoSharp.Core.SmartContract
+{
+    public enum TriggerType : byte
+    {
```
This enum is already on the nuget Package ( NeoSharp.VM.Interop )

shargon commented on src/NeoSharp.Core/Blockchain/Processing/InvocationTransactionPersister2.cs:
```diff
@@ -0,0 +1,31 @@
+namespace NeoSharp.Core.Blockchain.Processing
+{
+    public class InvocationTransactionPersister2: ITransactionPersister<InvocationTransaction>
+    {
```
Is this class used?

shargon commented on src/NeoSharp.Core/Blockchain/Processing/InvocationTransactionPersister.cs:
```diff
@@ -1,14 +1,74 @@
+
+        public InvocationTransactionPersister(
+            IVMFactory vmFactory,
+            IBinarySerializer binarySerializer,
```
We use the serializer is used in static way

shargon commented on src/NeoSharp.Core/VM/IStackAccessor.cs:
```diff
@@ -0,0 +1,35 @@
+namespace NeoSharp.Core.VM
+{
+    public interface IStackAccessor
+    {
```
Is this interface needed? Then should be moved to the `NeoSharp.VM.Interop` package

osmirnov replied:
I've realized it was probably not the best thing to introduce.

shargon commented on src/NeoSharp.Core/VM/StateMachine.cs:
```diff
@@ -0,0 +1,354 @@
+
+        private bool Contract_Create(IStackAccessor stack)
+        {
+            var script = stack.PopByteArray();
```
This method call `Dispose` to the base StackItem?

osmirnov replied:
I'll probably switch to VM abstractions instead. But yes it's a good point.

shargon commented on src/NeoSharp.VM.NeoVM/CryptoAdapter.cs:
```diff
@@ -0,0 +1,23 @@
+namespace NeoSharp.VM.NeoVM
+{
+    public class CryptoAdapter : ICrypto
+    {
```
We need this class?

shargon commented on src/NeoSharp.VM.NeoVM/IApplicationEngine.cs:
```diff
@@ -0,0 +1,25 @@
+namespace NeoSharp.VM.NeoVM
+{
+    public interface IApplicationEngine
+    {
```
Should implement the same interface between the VM, defined already in NeoSharp.VM.Interop

shargon reviewed (commented):
I think that both VM should share the same interfaces, defined in NeoSharp.VM.Interop.

shargon committed:
Try to unify with current interfaces

**Not tested**, first draft

This two PR could help with some problems

- https://github.com/neo-project/neo-vm/pull/64
- https://github.com/neo-project/neo-vm/pull/66

shargon commented:
@osmirnov there are problems with items like interopStackItem, i will deal with it on :

- https://github.com/neo-project/neo-vm/pull/64
- https://github.com/neo-project/neo-vm/pull/66

shargon committed:
Remove `this`

shargon commented:
https://github.com/neo-project/neo-vm/pull/67 this solution will help with this issues

shargon committed:
Workaround InteropInterface

shargon committed:
Syscall hashes

shargon committed:
Some fixes

shargon committed:
NeoVM.dll compiled (for advance without the last nuget)

shargon committed:
Merge branch 'development' into feat-initial-neo-vm-support

shargon commented:
@osmirnov please review the last changes, is using a compiled version of NeoVM with the last changes.

shargon committed:
Clean code

shargon committed:
Merge remote-tracking branch 'osmirnov/feat-initial-neo-vm-support' into feat-initial-neo-vm-support

shargon committed:
Clean code

shargon committed:
Clean code

shargon committed:
Free context

osmirnov committed:
change interop service

osmirnov committed:
convert abstract classes to base classes

osmirnov committed:
return null for ToByteArray when no conversion

shargon commented on src/NeoSharp.VM/Types/InteropService.cs:
```diff
@@ -88,84 +77,70 @@ public InteropService()
+                ? BitConverter.ToUInt32(method, 0)
+                : GetMethodHash(Encoding.ASCII.GetString(method));
+
+            void InvokeOnSysCall(SysCallResult result) => OnSysCall?.Invoke(this, new SysCallArgs(engine, GetMethodName(hash), hash, result));
```
This is very weird xD

osmirnov replied:
Yep, I used to use `Func` or `Action` delegates instead but VS converts them into local methods all the time.

shargon committed:
Small fix

shargon committed:
Test AVM

virtual contract add 0x0066
testinvoke 0x9e5bf7f291e4d8edd25e7b3c790c6f11590f9754 Application

shargon commented:
Execution successful of virtual smart contract

```
virtual contract add 0x0066
testinvoke 0x9e5bf7f291e4d8edd25e7b3c790c6f11590f9754 Application
```

![image](https://user-images.githubusercontent.com/3167973/49072782-51883380-f231-11e8-995c-9b65b192ab80.png)

shargon committed:
Clean code

shargon committed:
Centralize Gas increase

shargon committed:
Delegate SYSCALL gas to InteropService

shargon committed:
Our State is not a Flag

shargon committed:
Use nuGet package 2.3.3.1

shargon committed:
Fix the fix

shargon commented:
Now is used the nuGet package, Erik released this version :)

shargon commented:
@osmirnov do you know why travis is throwing an exception?

osmirnov commented on src/NeoSharp.VM.NeoVM/ApplicationEngine.cs:
```diff
@@ -0,0 +1,676 @@
+
+            _contractScriptTable = new ContractScriptTable(args.ScriptTable);
+
+            _engine = new ExecutionEngine(
```
@shargon it has to be disposed, right?

shargon replied:
Is from NeoVM, so it is not necessary

osmirnov commented:
> @osmirnov do you know why travis is throwing an exception?

Yes, I think it is related to https://github.com/CityOfZion/neo-sharp/pull/548/files#diff-432d787ad9965e43ec651da1be56928dR80 not to this PR

shargon committed:
Remove IDisposable from ExecutionContext and Stacks
